Trigger collection notifications when the source object is deleted

ironage opened the pull request:
This is an enhancement to fire notifications when an empty collection has been deleted because the underlying object was deleted. This was already working for the case where there were items in the collection, so this is just an enhancement to for people who would expect to be notified when the object itself is deleted. Even though it is a behavioural change, I don't think it is breaking because there are no changes reported in the notifier. It is up to the SDK's if they would like to somehow expose the new `collection_root_was_deleted` flag somehow if that is appropriate. Fixes #4228.

tgoyne commented on src/realm/object-store/impl/collection_change_builder.cpp:
```diff
@@ -25,10 +25,46 @@
 using namespace realm;
 using namespace realm::_impl;
 
+CollectionChangeBuilder::CollectionChangeBuilder(CollectionChangeBuilder&& other)
```
These shouldn't need to be overridden. I'm assuming it's because there's something which depends on collection_root_was_deleted being false in a moved-from CCB, but that's generally a bad thing to do and instead anything which cares about the state of a moved-from object should explicitly set it.

tgoyne commented on src/realm/object-store/impl/collection_change_builder.cpp:
```diff
@@ -675,6 +714,12 @@ CollectionChangeSet CollectionChangeBuilder::finalize() &&
     // but we don't want inserts in the final modification set
     modifications.remove(insertions);
 
+    // moving a bool will not reset it's value, so to get the move out behaviour
```
This should be done by the caller, not this. This function moves as an optimization and doesn't guarantee anything about the state of the object afterwards.

tgoyne reviewed (commented):
This is actually a breaking change (it'll currently result in us sending a `.initial` notification when the parent is deleted) but not a difficult one to adapt to so it's fine for Cocoa at least.

jedelbo reviewed (commented):
LGTM. I will leave it to @tgoyne to approve.

ironage commented:
I've updated the call sites where we relied on moving a changeset to clear it's state, thanks for pointing this out, it does feel much cleaner to do it this way. I've also moved the change note to the breaking section.
